File: express/src/openapi.js
/**
 * OpenAPI 3.0 spec for Express API (Bolt-compatible).
 * Used by Swagger UI (/docs) and ReDoc (/redoc).
 */

import { config } from "./config.js";

export function getOpenApiSpec(req) {
  let baseUrl = req ? `${req.protocol}://${req.get("host")}` : `http://localhost:${config.port}`;
  // 0.0.0.0 → localhost (brauzer "Try it out" uchun)
  if (baseUrl.includes("0.0.0.0")) {
    baseUrl = baseUrl.replace("0.0.0.0", "localhost");
  }

  return {
    openapi: "3.0.3",
    info: {
      title: "Express API (Bolt-compatible)",
      description:
        "Same endpoints as Django Bolt. Health, Roles, Users, Auth. JWT via POST /auth/login. Use **Authorize** to add Bearer token.",
      version: "0.1.0",
    },
    servers: [{ url: baseUrl, description: "API server" }],
    components: {
      securitySchemes: {
        BearerAuth: {
          type: "http",
          scheme: "bearer",
          bearerFormat: "JWT",
          description:
            "JWT from POST /auth/login. Click Authorize, enter: Bearer <your_access_token>",
        },
      },
      schemas: {
        HealthResponse: {
          type: "object",
          properties: { status: { type: "string", example: "ok" } },
        },
        HealthTestResponse: {
          type: "object",
          properties: {
            status: { type: "string", example: "ok" },
            message: { type: "string", example: "Test health check endpoint" },
          },
        },
        ReadyResponse: {
          type: "object",
          properties: {
            status: { type: "string", enum: ["healthy", "unhealthy"] },
            checks: {
              type: "object",
              properties: { database: { type: "string" } },
            },
          },
        },
        RoleSchema: {
          type: "object",
          properties: {
            code: { type: "string", example: "ADMIN" },
            name: { type: "string", example: "Administrator" },
          },
        },
        UserSchema: {
          type: "object",
          properties: {
            id: { type: "integer", example: 1 },
            username: { type: "string", example: "admin" },
            role: { type: "string", example: "ADMIN" },
          },
        },
        UserListResponse: {
          type: "object",
          properties: {
            results: { type: "array", items: { $ref: "#/components/schemas/UserSchema" } },
            count: { type: "integer" },
            next: { type: "string", nullable: true },
            previous: { type: "string", nullable: true },
          },
        },
        LoginSchema: {
          type: "object",
          required: ["username", "password"],
          properties: {
            username: { type: "string", example: "admin" },
            password: { type: "string", example: "password" },
          },
        },
        TokenSchema: {
          type: "object",
          properties: {
            access_token: { type: "string" },
            expires_in: { type: "integer" },
            token_type: { type: "string", example: "bearer" },
          },
        },
        ErrorDetail: {
          type: "object",
          properties: { detail: { type: "string" } },
        },
      },
    },
    paths: {
      "/health": {
        get: {
          tags: ["Health"],
          summary: "Liveness",
          description: "Simple liveness check.",
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/HealthResponse" },
                },
              },
            },
          },
        },
      },
      "/health/test": {
        get: {
          tags: ["Health"],
          summary: "Custom health check",
          description: "Test health check endpoint (Bolt-compatible).",
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/HealthTestResponse" },
                },
              },
            },
          },
        },
      },
      "/ready": {
        get: {
          tags: ["Health"],
          summary: "Readiness",
          description: "Readiness check (DB). Returns 503 if unhealthy.",
          responses: {
            "200": {
              description: "Healthy",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ReadyResponse" },
                },
              },
            },
            "503": {
              description: "Unhealthy",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ReadyResponse" },
                },
              },
            },
          },
        },
      },
      "/roles": {
        get: {
          tags: ["Roles"],
          summary: "List roles",
          description: "List all roles (ADMIN, SHOPKEEPER, CUSTOMER).",
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: {
                    type: "array",
                    items: { $ref: "#/components/schemas/RoleSchema" },
                  },
                },
              },
            },
          },
        },
      },
      "/roles/code/{code}": {
        get: {
          tags: ["Roles"],
          summary: "Get role by code",
          parameters: [
            {
              name: "code",
              in: "path",
              required: true,
              schema: { type: "string", example: "ADMIN" },
            },
          ],
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/RoleSchema" },
                },
              },
            },
            "404": {
              description: "Role not found",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ErrorDetail" },
                },
              },
            },
          },
        },
      },
      "/users": {
        get: {
          tags: ["Users"],
          summary: "List users",
          description: "Paginated list with search and role filter.",
          parameters: [
            { name: "search", in: "query", schema: { type: "string" } },
            { name: "role", in: "query", schema: { type: "string", enum: ["ADMIN", "SHOPKEEPER", "CUSTOMER"] } },
            { name: "role_code", in: "query", schema: { type: "string" } },
            { name: "page", in: "query", schema: { type: "integer", default: 1 } },
            { name: "page_size", in: "query", schema: { type: "integer", default: 10 } },
          ],
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/UserListResponse" },
                },
              },
            },
          },
        },
      },
      "/users/{id}": {
        get: {
          tags: ["Users"],
          summary: "Get user by ID",
          parameters: [
            {
              name: "id",
              in: "path",
              required: true,
              schema: { type: "integer" },
            },
          ],
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/UserSchema" },
                },
              },
            },
            "404": {
              description: "User not found",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ErrorDetail" },
                },
              },
            },
          },
        },
      },
      "/auth/login": {
        post: {
          tags: ["Auth"],
          summary: "Login (JWT)",
          description: "Obtain JWT access token. No auth required.",
          requestBody: {
            required: true,
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/LoginSchema" },
              },
            },
          },
          responses: {
            "200": {
              description: "OK",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/TokenSchema" },
                },
              },
            },
            "400": {
              description: "Bad request",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ErrorDetail" },
                },
              },
            },
            "401": {
              description: "Invalid credentials",
              content: {
                "application/json": {
                  schema: { $ref: "#/components/schemas/ErrorDetail" },
                },
              },
            },
          },
        },
      },
    },
  };
}

File: go/internal/docs/openapi.go
package docs

import (
	"encoding/json"
	"net/http"
)

// OpenAPI returns the OpenAPI 3.0 spec for the Go API (Bolt-compatible).
// All responses include X-Server-Time and X-Response-Time headers.
func OpenAPI(baseURL string) map[string]any {
	if baseURL == "" {
		baseURL = "http://localhost:8005"
	}
	return map[string]any{
		"openapi": "3.0.3",
		"info": map[string]any{
			"title":       "Go API (Bolt-compatible)",
			"description": "Same endpoints as Django Bolt. Health, Roles, Users, Auth. JWT via POST /auth/login. All responses include **X-Response-Time** header. Use **Authorize** to add Bearer token.",
			"version":     "0.1.0",
		},
		"servers": []map[string]any{
			{"url": baseURL, "description": "API server"},
		},
		"components": map[string]any{
			"securitySchemes": map[string]any{
				"BearerAuth": map[string]any{
					"type":         "http",
					"scheme":       "bearer",
					"bearerFormat": "JWT",
					"description":  "JWT from POST /auth/login. Click Authorize, enter: Bearer <your_access_token>",
				},
			},
			"schemas": map[string]any{
				"HealthResponse": map[string]any{
					"type": "object",
					"properties": map[string]any{"status": map[string]any{"type": "string", "example": "ok"}},
				},
				"HealthTestResponse": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"status":  map[string]any{"type": "string", "example": "ok"},
						"message": map[string]any{"type": "string", "example": "Test health check endpoint"},
					},
				},
				"ReadyResponse": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"status": map[string]any{"type": "string", "enum": []string{"healthy", "unhealthy"}},
						"checks": map[string]any{
							"type": "object",
							"properties": map[string]any{"database": map[string]any{"type": "string"}},
						},
					},
				},
				"RoleSchema": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"code": map[string]any{"type": "string", "example": "ADMIN"},
						"name": map[string]any{"type": "string", "example": "Administrator"},
					},
				},
				"UserSchema": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"id":       map[string]any{"type": "integer", "example": 1},
						"username": map[string]any{"type": "string", "example": "admin"},
						"role":     map[string]any{"type": "string", "example": "ADMIN"},
					},
				},
				"UserListResponse": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"results":  map[string]any{"type": "array", "items": map[string]any{"$ref": "#/components/schemas/UserSchema"}},
						"count":    map[string]any{"type": "integer"},
						"next":     map[string]any{"type": "string", "nullable": true},
						"previous": map[string]any{"type": "string", "nullable": true},
					},
				},
				"LoginSchema": map[string]any{
					"type":     "object",
					"required": []string{"username", "password"},
					"properties": map[string]any{
						"username": map[string]any{"type": "string", "example": "admin"},
						"password": map[string]any{"type": "string", "example": "password"},
					},
				},
				"TokenSchema": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"access_token": map[string]any{"type": "string"},
						"expires_in":   map[string]any{"type": "integer"},
						"token_type":   map[string]any{"type": "string", "example": "bearer"},
					},
				},
				"ErrorDetail": map[string]any{
					"type":       "object",
					"properties": map[string]any{"detail": map[string]any{"type": "string"}},
				},
			},
		},
		"paths": map[string]any{
			"/health": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Health"},
					"summary":     "Liveness",
					"description": "Simple liveness check. Response includes X-Response-Time header.",
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"headers": map[string]any{
								"X-Response-Time": map[string]any{"schema": map[string]any{"type": "string"}, "description": "Request processing time (ms)"},
								"X-Server-Time":   map[string]any{"schema": map[string]any{"type": "string"}, "description": "UTC timestamp"},
							},
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/HealthResponse"},
								},
							},
						},
					},
				},
			},
			"/health/test": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Health"},
					"summary":     "Custom health check",
					"description": "Test health check endpoint (Bolt-compatible).",
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/HealthTestResponse"},
								},
							},
						},
					},
				},
			},
			"/ready": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Health"},
					"summary":     "Readiness",
					"description": "Readiness check (DB). Returns 503 if unhealthy.",
					"responses": map[string]any{
						"200": map[string]any{
							"description": "Healthy",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ReadyResponse"},
								},
							},
						},
						"503": map[string]any{
							"description": "Unhealthy",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ReadyResponse"},
								},
							},
						},
					},
				},
			},
			"/roles": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Roles"},
					"summary":     "List roles",
					"description": "List all roles (ADMIN, SHOPKEEPER, CUSTOMER).",
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{
										"type":  "array",
										"items": map[string]any{"$ref": "#/components/schemas/RoleSchema"},
									},
								},
							},
						},
					},
				},
			},
			"/roles/code/{code}": map[string]any{
				"get": map[string]any{
					"tags": []string{"Roles"},
					"summary": "Get role by code",
					"parameters": []map[string]any{
						{"name": "code", "in": "path", "required": true, "schema": map[string]any{"type": "string", "example": "ADMIN"}},
					},
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/RoleSchema"},
								},
							},
						},
						"404": map[string]any{
							"description": "Role not found",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
					},
				},
			},
			"/users": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Users"},
					"summary":     "List users",
					"description": "Paginated list with search and role filter.",
					"parameters": []map[string]any{
						{"name": "search", "in": "query", "schema": map[string]any{"type": "string"}},
						{"name": "role", "in": "query", "schema": map[string]any{"type": "string", "enum": []string{"ADMIN", "SHOPKEEPER", "CUSTOMER"}}},
						{"name": "role_code", "in": "query", "schema": map[string]any{"type": "string"}},
						{"name": "page", "in": "query", "schema": map[string]any{"type": "integer", "default": 1}},
						{"name": "page_size", "in": "query", "schema": map[string]any{"type": "integer", "default": 10}},
					},
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/UserListResponse"},
								},
							},
						},
					},
				},
				"post": map[string]any{
					"tags":        []string{"Users"},
					"summary":     "Create user",
					"description": "Create user (staff only). Requires JWT.",
					"security":    []map[string]any{{"BearerAuth": []string{}}},
					"requestBody": map[string]any{
						"required": true,
						"content": map[string]any{
							"application/json": map[string]any{
								"schema": map[string]any{
									"type":     "object",
									"required": []string{"username", "password"},
									"properties": map[string]any{
										"username": map[string]any{"type": "string"},
										"password": map[string]any{"type": "string"},
										"email":    map[string]any{"type": "string"},
										"role":     map[string]any{"type": "string", "enum": []string{"ADMIN", "SHOPKEEPER", "CUSTOMER"}},
									},
								},
							},
						},
					},
					"responses": map[string]any{
						"201": map[string]any{
							"description": "Created",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/UserSchema"},
								},
							},
						},
						"400": map[string]any{
							"description": "Bad request",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
						"401": map[string]any{
							"description": "Unauthorized",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
					},
				},
			},
			"/users/{user_id}": map[string]any{
				"get": map[string]any{
					"tags": []string{"Users"},
					"summary": "Get user by ID",
					"parameters": []map[string]any{
						{"name": "user_id", "in": "path", "required": true, "schema": map[string]any{"type": "integer"}},
					},
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/UserSchema"},
								},
							},
						},
						"404": map[string]any{
							"description": "User not found",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
					},
				},
			},
			"/users/me": map[string]any{
				"get": map[string]any{
					"tags":        []string{"Users"},
					"summary":     "Current user",
					"description": "Get current authenticated user. Requires JWT.",
					"security":    []map[string]any{{"BearerAuth": []string{}}},
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/UserSchema"},
								},
							},
						},
						"401": map[string]any{
							"description": "Unauthorized",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
					},
				},
			},
			"/auth/login": map[string]any{
				"post": map[string]any{
					"tags":        []string{"Auth"},
					"summary":     "Login (JWT)",
					"description": "Obtain JWT access token. No auth required.",
					"requestBody": map[string]any{
						"required": true,
						"content": map[string]any{
							"application/json": map[string]any{
								"schema": map[string]any{"$ref": "#/components/schemas/LoginSchema"},
							},
						},
					},
					"responses": map[string]any{
						"200": map[string]any{
							"description": "OK",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/TokenSchema"},
								},
							},
						},
						"400": map[string]any{
							"description": "Bad request",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
						"401": map[string]any{
							"description": "Invalid credentials",
							"content": map[string]any{
								"application/json": map[string]any{
									"schema": map[string]any{"$ref": "#/components/schemas/ErrorDetail"},
								},
							},
						},
					},
				},
			},
		},
	}
}

// ServeOpenAPI returns a handler that serves the OpenAPI spec as JSON.
// baseURL: e.g. "http://localhost:8005". If empty, uses request Host.
func ServeOpenAPI(defaultBaseURL string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		baseURL := defaultBaseURL
		if baseURL == "" && r.Host != "" {
			scheme := "http"
			if r.TLS != nil {
				scheme = "https"
			}
			baseURL = scheme + "://" + r.Host
		}
		if baseURL == "" {
			baseURL = "http://localhost:8005"
		}
		spec := OpenAPI(baseURL)
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(spec)
	}
}
